sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/udf/BuiltinStringFunctions.java

```diff
@@ -0,0 +1,302 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.impl.udf;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.google.auto.service.AutoService;
+import java.util.Arrays;
+import org.apache.beam.sdk.schemas.Schema.TypeName;
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+
+/** BuiltinStringFunctions. */
+@AutoService(BeamBuiltinFunctionProvider.class)
+public class BuiltinStringFunctions extends BeamBuiltinFunctionProvider {
+
+  // return a explicitly null for Boolean has NP_BOOLEAN_RETURN_NULL warning.
+  // return null for boolean is not allowed.
+  // TODO: handle null input.
+  @UDF(
+    funcName = "ENDS_WITH",
+    parameterArray = {TypeName.STRING},
+    returnType = TypeName.STRING
+  )
+  public Boolean endsWith(String str1, String str2) {
+    return str1.endsWith(str2);
+  }
+
+  // return a explicitly null for Boolean has NP_BOOLEAN_RETURN_NULL warning.
+  // return null for boolean is not allowed.
+  // TODO: handle null input.
+  @UDF(
+    funcName = "STARTS_WITH",
+    parameterArray = {TypeName.STRING},
+    returnType = TypeName.STRING
+  )
+  public Boolean startsWith(String str1, String str2) {
+    return str1.startsWith(str2);
+  }
+
+  @UDF(
+    funcName = "LENGTH",
```

Comment:
This looks to be identical to CHARACTER_LENGTH in calcite?

Reply:
As the reference, [functions-and-operators#length](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#length), `length` handles `character array` and `string`. In calcite, CHARACTER_LENGTH only handle `string` (https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java#L235).


In BigQuery, `string` is defined as UTF8 encoded character array. Functions handle string and character array differently. See: https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#string-type

Reply:
You can go ahead to use code generation. For any function that we might want them behave differently, we can implement those and let planner decides which set of functions to use.